File: webapp/src/templates/templates.ts
// Sample configuration templates for different hydraulic network scenarios

export interface Template {
  id: string;
  name: string;
  description: string;
  category: 'basic' | 'industrial' | 'gas' | 'liquid' | 'advanced' | 'custom';
  complexity: 'simple' | 'medium' | 'complex';
  tags: string[];
  content: any;
  createdAt: Date;
  updatedAt: Date;
  author?: string;
  version?: string;
  isFeatured?: boolean;
  estimatedTime?: string;
}

// Basic Liquid System Template
export const basicLiquidSystem: Template = {
  id: 'basic-liquid-001',
  name: 'Basic Liquid Pipeline',
  description: 'Simple liquid pipeline with single phase flow',
  category: 'basic',
  complexity: 'simple',
  tags: ['liquid', 'pipeline', 'basic', 'single-phase'],
  estimatedTime: '2-5 minutes',
  author: 'System',
  version: '1.0',
  isFeatured: true,
  createdAt: new Date('2024-01-01'),
  updatedAt: new Date('2024-01-01'),
  content: {
    network: {
      name: 'Basic Liquid Pipeline',
      description: 'Simple pipeline system for liquid transport',
      direction: 'forward',
      boundaryPressure: 200000,
      outputUnits: {
        pressure: 'kPa',
        temperature: 'C',
        length: 'm',
        flowRate: 'L/min'
      },
      designMargin: 10
    },
    fluid: {
      name: 'Water',
      phase: 'liquid',
      temperature: 293.15, // 20°C
      pressure: 200000, // 200 kPa
      density: 998, // kg/m³
      viscosity: 0.001, // Pa·s
      massFlowRate: 10 // kg/s
    },
    sections: [
      {
        id: 'section-1',
        description: 'Main pipeline section',
        schedule: '40',
        pipeNPD: 4, // 4 inch NPS
        roughness: 4.6e-5, // m (steel pipe)
        length: 100, // m
        elevationChange: 10, // m
        fittingType: 'LR',
        fittings: [
          {
            type: '90-ell',
            count: 4,
            kFactor: 0.9
          },
          {
            type: 'gate-valve',
            count: 2,
            kFactor: 0.17
          }
        ]
      }
    ]
  }
};

// Gas Transmission System Template
export const gasTransmissionSystem: Template = {
  id: 'gas-transmission-001',
  name: 'Gas Transmission Pipeline',
  description: 'High-pressure gas transmission system with compressibility effects',
  category: 'gas',
  complexity: 'medium',
  tags: ['gas', 'transmission', 'high-pressure', 'compressible'],
  estimatedTime: '5-10 minutes',
  author: 'System',
  version: '1.0',
  isFeatured: true,
  createdAt: new Date('2024-01-01'),
  updatedAt: new Date('2024-01-01'),
  content: {
    network: {
      name: 'Gas Transmission Pipeline',
      description: 'High-pressure natural gas transmission system',
      direction: 'forward',
      upstreamPressure: 8000000, // 8 MPa
      downstreamPressure: 5000000, // 5 MPa
      gasFlowModel: 'isothermal',
      outputUnits: {
        pressure: 'MPa',
        temperature: 'C',
        length: 'km',
        flowRate: 'MMSCFD'
      }
    },
    fluid: {
      name: 'Natural Gas',
      phase: 'gas',
      temperature: 303.15, // 30°C
      pressure: 8000000, // 8 MPa
      molecularWeight: 18.5, // kg/kmol
      zFactor: 0.92,
      specificHeatRatio: 1.28,
      viscosity: 1.2e-5, // Pa·s
      standardFlowRate: 50 // MMSCFD
    },
    sections: [
      {
        id: 'section-1',
        description: 'Main transmission line',
        schedule: '40',
        pipeNPD: 24, // 24 inch NPS
        roughness: 4.6e-5, // m
        length: 50000, // m
        elevationChange: 50, // m
        fittingType: 'LR',
        fittings: [
          {
            type: 'welded-ell',
            count: 8,
            kFactor: 0.5
          },
          {
            type: 'ball-valve',
            count: 3,
            kFactor: 0.1
          }
        ]
      },
      {
        id: 'section-2',
        description: 'Reducing section',
        schedule: '40',
        pipeNPD: 16, // 16 inch NPS
        roughness: 4.6e-5,
        length: 10000, // m
        elevationChange: 20, // m
        fittingType: 'SR',
        fittings: [
          {
            type: 'concentric-reducer',
            count: 1,
            kFactor: 0.05
          }
        ]
      }
    ]
  }
};

// Industrial Process System Template
export const industrialProcessSystem: Template = {
  id: 'industrial-process-001',
  name: 'Industrial Process System',
  description: 'Complex industrial process with multiple branches and equipment',
  category: 'industrial',
  complexity: 'complex',
  tags: ['industrial', 'process', 'branching', 'equipment'],
  estimatedTime: '10-20 minutes',
  author: 'System',
  version: '1.0',
  isFeatured: false,
  createdAt: new Date('2024-01-01'),
  updatedAt: new Date('2024-01-01'),
  content: {
    network: {
      name: 'Industrial Process System',
      description: 'Multi-branch industrial process with pumps and heat exchangers',
      direction: 'auto',
      boundaryPressure: 300000,
      outputUnits: {
        pressure: 'bar',
        temperature: 'C',
        length: 'm',
        flowRate: 'm3/h'
      }
    },
    fluid: {
      name: 'Process Fluid',
      phase: 'liquid',
      temperature: 350, // 77°C
      pressure: 300000, // 3 bar
      density: 850, // kg/m³
      viscosity: 0.002, // Pa·s
      massFlowRate: 50 // kg/s
    },
    sections: [
      {
        id: 'section-1',
        description: 'Main process line',
        schedule: '80',
        pipeNPD: 6, // 6 inch NPS
        roughness: 4.6e-5,
        length: 50, // m
        elevationChange: 0, // m
        fittingType: 'LR',
        fittings: [
          {
            type: '90-ell',
            count: 6,
            kFactor: 0.9
          },
          {
            type: 'check-valve',
            count: 1,
            kFactor: 2.0
          }
        ],
        controlValve: {
          type: 'globe-valve',
          coefficient: 50,
          opening: 0.75
        }
      },
      {
        id: 'section-2',
        description: 'Heat exchanger bypass',
        schedule: '40',
        pipeNPD: 4, // 4 inch NPS
        roughness: 4.6e-5,
        length: 30, // m
        elevationChange: 5, // m
        fittingType: 'SR',
        fittings: [
          {
            type: '45-ell',
            count: 4,
            kFactor: 0.4
          }
        ],
        orifice: {
          diameter: 0.05, // m
          coefficient: 0.62
        }
      }
    ]
  }
};

// Cryogenic System Template
export const cryogenicSystem: Template = {
  id: 'cryogenic-001',
  name: 'Cryogenic Liquid System',
  description: 'Low temperature liquid system with special considerations',
  category: 'advanced',
  complexity: 'complex',
  tags: ['cryogenic', 'low-temperature', 'liquid', 'specialized'],
  estimatedTime: '15-25 minutes',
  author: 'System',
  version: '1.0',
  isFeatured: false,
  createdAt: new Date('2024-01-01'),
  updatedAt: new Date('2024-01-01'),
  content: {
    network: {
      name: 'Cryogenic Liquid System',
      description: 'Liquid nitrogen distribution system with insulation considerations',
      direction: 'forward',
      upstreamPressure: 500000, // 5 bar
      downstreamPressure: 300000, // 3 bar
      outputUnits: {
        pressure: 'bar',
        temperature: 'K',
        length: 'm',
        flowRate: 'kg/s'
      }
    },
    fluid: {
      name: 'Liquid Nitrogen',
      phase: 'liquid',
      temperature: 77.36, // Boiling point at 1 atm
      pressure: 500000, // 5 bar
      density: 807, // kg/m³
      viscosity: 1.56e-4, // Pa·s
      massFlowRate: 5 // kg/s
    },
    sections: [
      {
        id: 'section-1',
        description: 'Insulated supply line',
        schedule: '40',
        pipeNPD: 2, // 2 inch NPS
        roughness: 1.5e-6, // m (polished stainless steel)
        length: 100, // m
        elevationChange: 20, // m
        fittingType: 'LR',
        fittings: [
          {
            type: 'sanitary-ell',
            count: 8,
            kFactor: 0.7
          },
          {
            type: 'diaphragm-valve',
            count: 2,
            kFactor: 1.5
          }
        ]
      },
      {
        id: 'section-2',
        description: 'Distribution manifold',
        schedule: '40',
        pipeNPD: 1, // 1 inch NPS
        roughness: 1.5e-6,
        length: 25, // m
        elevationChange: 5, // m
        fittingType: 'SR',
        fittings: [
          {
            type: 'sanitary-tee',
            count: 3,
            kFactor: 2.0
          }
        ]
      }
    ]
  }
};

// Simple Gas Vent System Template
export const simpleGasVent: Template = {
  id: 'simple-gas-vent-001',
  name: 'Simple Gas Vent System',
  description: 'Basic atmospheric vent system for gas release',
  category: 'basic',
  complexity: 'simple',
  tags: ['gas', 'vent', 'atmospheric', 'safety'],
  estimatedTime: '2-5 minutes',
  author: 'System',
  version: '1.0',
  isFeatured: false,
  createdAt: new Date('2024-01-01'),
  updatedAt: new Date('2024-01-01'),
  content: {
    network: {
      name: 'Gas Vent System',
      description: 'Simple vent to atmosphere',
      direction: 'forward',
      upstreamPressure: 200000, // 2 bar
      downstreamPressure: 101325, // Atmospheric
      gasFlowModel: 'adiabatic',
      outputUnits: {
        pressure: 'kPa',
        temperature: 'C',
        length: 'm',
        flowRate: 'm3/h'
      }
    },
    fluid: {
      name: 'Process Gas',
      phase: 'gas',
      temperature: 300, // 27°C
      pressure: 200000, // 2 bar
      molecularWeight: 28.97, // Air
      zFactor: 1.0,
      specificHeatRatio: 1.4,
      viscosity: 1.8e-5, // Pa·s
      volumetricFlowRate: 100 // m3/h
    },
    sections: [
      {
        id: 'section-1',
        description: 'Vent pipe',
        schedule: '40',
        pipeNPD: 2, // 2 inch NPS
        roughness: 4.6e-5,
        length: 10, // m
        elevationChange: 5, // m
        fittingType: 'SR',
        fittings: [
          {
            type: '90-ell',
            count: 2,
            kFactor: 0.9
          },
          {
            type: 'vent-stack',
            count: 1,
            kFactor: 1.0
          }
        ]
      }
    ]
  }
};

// Export all templates
export const templates: Template[] = [
  basicLiquidSystem,
  gasTransmissionSystem,
  industrialProcessSystem,
  cryogenicSystem,
  simpleGasVent
];

// Template categories for filtering
export const templateCategories = [
  { value: 'all', label: 'All Templates', count: templates.length },
  { value: 'basic', label: 'Basic', count: templates.filter(t => t.category === 'basic').length },
  { value: 'liquid', label: 'Liquid Systems', count: templates.filter(t => t.category === 'liquid').length },
  { value: 'gas', label: 'Gas Systems', count: templates.filter(t => t.category === 'gas').length },
  { value: 'industrial', label: 'Industrial', count: templates.filter(t => t.category === 'industrial').length },
  { value: 'advanced', label: 'Advanced', count: templates.filter(t => t.category === 'advanced').length }
];

// Get template by ID
export const getTemplateById = (id: string): Template | undefined => {
  return templates.find(template => template.id === id);
};

// Get templates by category
export const getTemplatesByCategory = (category: string): Template[] => {
  if (category === 'all') return templates;
  return templates.filter(template => template.category === category);
};

// Get featured templates
export const getFeaturedTemplates = (): Template[] => {
  return templates.filter(template => template.isFeatured);
};

// Get templates by complexity
export const getTemplatesByComplexity = (complexity: string): Template[] => {
  return templates.filter(template => template.complexity === complexity);
};

// Search templates by tags
export const searchTemplates = (query: string): Template[] => {
  const lowerQuery = query.toLowerCase();
  return templates.filter(template => 
    template.name.toLowerCase().includes(lowerQuery) ||
    template.description.toLowerCase().includes(lowerQuery) ||
    template.tags.some(tag => tag.toLowerCase().includes(lowerQuery))
  );
};

File: backend/templates.py
"""Configuration templates for common hydraulic network scenarios.

This module provides sample configurations for various hydraulic network
scenarios including liquid and gas systems, different pipe configurations,
and common industrial applications.
"""

from typing import Any, Dict, List


def get_liquid_system_template() -> Dict[str, Any]:
    """Get a template for liquid (water) system calculations.
    
    Returns:
        Configuration dictionary for liquid system
    """
    return {
        "network": {
            "name": "Liquid System Example",
            "description": "Example liquid (water) system with multiple pipe sections",
            "direction": "forward",
            "boundary_pressure": {
                "value": 200.0,
                "unit": "kPag"
            },
            "design_margin": 10.0,
            "mass_flow_rate": {
                "value": 1000.0,
                "unit": "kg/h"
            },
            "output_units": {
                "pressure": "kPag",
                "pressure_drop": "kPa",
                "temperature": "degC",
                "density": "kg/m^3",
                "velocity": "m/s",
                "volumetric_flow_rate": "m^3/h",
                "mass_flow_rate": "kg/h"
            },
            "fluid": {
                "name": "Water",
                "phase": "liquid",
                "temperature": {
                    "value": 25.0,
                    "unit": "degC"
                },
                "pressure": {
                    "value": 200.0,
                    "unit": "kPag"
                },
                "density": {
                    "value": 998.0,
                    "unit": "kg/m^3"
                },
                "viscosity": {
                    "value": 1.002,
                    "unit": "cP"
                }
            },
            "sections": [
                {
                    "id": "inlet",
                    "description": "Inlet pipe section",
                    "schedule": "40",
                    "pipe_NPD": 4.0,
                    "roughness": 4.57e-5,
                    "length": {
                        "value": 10.0,
                        "unit": "m"
                    },
                    "elevation_change": {
                        "value": 0.0,
                        "unit": "m"
                    },
                    "erosional_constant": 100,
                    "fitting_type": "LR",
                    "fittings": [
                        {"type": "pipe_entrance_normal", "count": 1},
                        {"type": "elbow_90", "count": 2}
                    ]
                },
                {
                    "id": "main_line",
                    "description": "Main pipeline section",
                    "schedule": "40",
                    "pipe_NPD": 4.0,
                    "roughness": 4.57e-5,
                    "length": {
                        "value": 50.0,
                        "unit": "m"
                    },
                    "elevation_change": {
                        "value": 5.0,
                        "unit": "m"
                    },
                    "erosional_constant": 100,
                    "fitting_type": "LR",
                    "fittings": [
                        {"type": "elbow_90", "count": 3},
                        {"type": "tee_through", "count": 1}
                    ]
                },
                {
                    "id": "outlet",
                    "description": "Outlet pipe section",
                    "schedule": "40",
                    "pipe_NPD": 4.0,
                    "roughness": 4.57e-5,
                    "length": {
                        "value": 15.0,
                        "unit": "m"
                    },
                    "elevation_change": {
                        "value": 2.0,
                        "unit": "m"
                    },
                    "erosional_constant": 100,
                    "fitting_type": "LR",
                    "fittings": [
                        {"type": "elbow_90", "count": 2},
                        {"type": "pipe_exit", "count": 1}
                    ]
                }
            ]
        }
    }


def get_gas_system_template() -> Dict[str, Any]:
    """Get a template for gas (natural gas) system calculations.
    
    Returns:
        Configuration dictionary for gas system
    """
    return {
        "network": {
            "name": "Gas System Example",
            "description": "Example gas (natural gas) system with isothermal flow",
            "direction": "forward",
            "boundary_pressure": {
                "value": 500.0,
                "unit": "psig"
            },
            "gas_flow_model": "isothermal",
            "design_margin": 5.0,
            "volumetric_flow_rate": {
                "value": 1000.0,
                "unit": "scfh"
            },
            "output_units": {
                "pressure": "psig",
                "pressure_drop": "psi",
                "temperature": "degF",
                "density": "lb/ft^3",
                "velocity": "ft/s",
                "volumetric_flow_rate": "scfh",
                "mass_flow_rate": "lb/h"
            },
            "fluid": {
                "name": "Natural Gas",
                "phase": "gas",
                "temperature": {
                    "value": 60.0,
                    "unit": "degF"
                },
                "pressure": {
                    "value": 500.0,
                    "unit": "psig"
                },
                "molecular_weight": {
                    "value": 18.0,
                    "unit": "lb/lbmol"
                },
                "z_factor": 0.95,
                "specific_heat_ratio": 1.30,
                "viscosity": {
                    "value": 0.012,
                    "unit": "cP"
                }
            },
            "sections": [
                {
                    "id": "compressor_discharge",
                    "description": "Compressor discharge line",
                    "schedule": "40",
                    "pipe_NPD": 6.0,
                    "roughness": 4.57e-5,
                    "length": {
                        "value": 20.0,
                        "unit": "m"
                    },
                    "elevation_change": {
                        "value": 0.0,
                        "unit": "m"
                    },
                    "erosional_constant": 100,
                    "fitting_type": "LR",
                    "fittings": [
                        {"type": "elbow_90", "count": 4},
                        {"type": "check_valve_swing", "count": 1}
                    ]
                },
                {
                    "id": "distribution_line",
                    "description": "Main distribution line",
                    "schedule": "40",
                    "pipe_NPD": 6.0,
                    "roughness": 4.57e-5,
                    "length": {
                        "value": 100.0,
                        "unit": "m"
                    },
                    "elevation_change": {
                        "value": 10.0,
                        "unit": "m"
                    },
                    "erosional_constant": 100,
                    "fitting_type": "LR",
                    "fittings": [
                        {"type": "elbow_90", "count": 6},
                        {"type": "tee_through", "count": 2}
                    ]
                },
                {
                    "id": "delivery_line",
                    "description": "Final delivery line",
                    "schedule": "40",
                    "pipe_NPD": 4.0,
                    "roughness": 4.57e-5,
                    "length": {
                        "value": 30.0,
                        "unit": "m"
                    },
                    "elevation_change": {
                        "value": 5.0,
                        "unit": "m"
                    },
                    "erosional_constant": 100,
                    "fitting_type": "LR",
                    "fittings": [
                        {"type": "elbow_90", "count": 3},
                        {"type": "control_valve", "count": 1},
                        {"type": "pipe_exit", "count": 1}
                    ],
                    "control_valve": {
                        "tag": "CV-101",
                        "cv": 250.0,
                        "FL": 0.9,
                        "xT": 0.70
                    }
                }
            ]
        }
    }


def get_vapor_system_template() -> Dict[str, Any]:
    """Get a template for vapor (steam) system calculations.
    
    Returns:
        Configuration dictionary for vapor system
    """
    return {
        "network": {
            "name": "Steam System Example",
            "description": "Example steam system with adiabatic flow",
            "direction": "forward",
            "boundary_pressure": {
                "value": 150.0,
                "unit": "psig"
            },
            "gas_flow_model": "adiabatic",
            "design_margin": 15.0,
            "mass_flow_rate": {
                "value": 5000.0,
                "unit": "lb/h"
            },
            "output_units": {
                "pressure": "psig",
                "pressure_drop": "psi",
                "temperature": "degF",
                "density": "lb/ft^3",
                "velocity": "ft/s",
                "volumetric_flow_rate": "acfh",
                "mass_flow_rate": "lb/h"
            },
            "fluid": {
                "name": "Steam",
                "phase": "vapor",
                "temperature": {
                    "value": 366.0,
                    "unit": "degF"
                },
                "pressure": {
                    "value": 150.0,
                    "unit": "psig"
                },
                "molecular_weight": {
                    "value": 18.02,
                    "unit": "lb/lbmol"
                },
                "z_factor": 1.0,
                "specific_heat_ratio": 1.33,
                "viscosity": {
                    "value": 0.013,
                    "unit": "cP"
                }
            },
            "sections": [
                {
                    "id": "steam_line",
                    "description": "Main steam header",
                    "schedule": "40",
                    "pipe_NPD": 8.0,
                    "roughness": 4.57e-5,
                    "length": {
                        "value": 50.0,
                        "unit": "m"
                    },
                    "elevation_change": {
                        "value": 0.0,
                        "unit": "m"
                    },
                    "erosional_constant": 100,
                    "fitting_type": "LR",
                    "fittings": [
                        {"type": "elbow_90", "count": 5},
                        {"type": "tee_through", "count": 1}
                    ]
                },
                {
                    "id": "branch_line",
                    "description": "Branch line to equipment",
                    "schedule": "40",
                    "pipe_NPD": 4.0,
                    "roughness": 4.57e-5,
                    "length": {
                        "value": 25.0,
                        "unit": "m"
                    },
                    "elevation_change": {
                        "value": 8.0,
                        "unit": "m"
                    },
                    "erosional_constant": 100,
                    "fitting_type": "LR",
                    "fittings": [
                        {"type": "elbow_90", "count": 4},
                        {"type": "block_valve_full_line_size", "count": 1}
                    ],
                    "user_specified_fixed_loss": {
                        "value": 2.0,
                        "unit": "psi"
                    }
                }
            ]
        }
    }


def get_simple_liquid_template() -> Dict[str, Any]:
    """Get a simple liquid system template for testing.
    
    Returns:
        Simple configuration dictionary
    """
    return {
        "network": {
            "name": "Simple Liquid Test",
            "description": "Simple liquid system for testing",
            "direction": "forward",
            "boundary_pressure": {
                "value": 101.325,
                "unit": "kPa"
            },
            "mass_flow_rate": {
                "value": 100.0,
                "unit": "kg/h"
            },
            "fluid": {
                "name": "Water",
                "phase": "liquid",
                "temperature": {
                    "value": 20.0,
                    "unit": "degC"
                },
                "pressure": {
                    "value": 101.325,
                    "unit": "kPa"
                },
                "density": {
                    "value": 998.0,
                    "unit": "kg/m^3"
                },
                "viscosity": {
                    "value": 1.0,
                    "unit": "cP"
                }
            },
            "sections": [
                {
                    "id": "test_section",
                    "description": "Single test section",
                    "schedule": "40",
                    "pipe_NPD": 2.0,
                    "roughness": 4.57e-5,
                    "length": {
                        "value": 10.0,
                        "unit": "m"
                    },
                    "elevation_change": {
                        "value": 0.0,
                        "unit": "m"
                    },
                    "fitting_type": "LR",
                    "fittings": [
                        {"type": "elbow_90", "count": 2}
                    ]
                }
            ]
        }
    }


def get_fitting_library_data() -> List[Dict[str, Any]]:
    """Get standard fitting library data with K-factors.
    
    Returns:
        List of fitting property dictionaries
    """
    return [
        {
            "type": "elbow_90",
            "description": "90-degree standard elbow",
            "typical_k_factor": "0.75-1.2",
            "manufacturer_data": {
                "standard": 0.9,
                "long_radius": 0.5,
                "short_radius": 1.2
            },
            "reference": "Crane Technical Paper No. 410"
        },
        {
            "type": "elbow_45",
            "description": "45-degree standard elbow",
            "typical_k_factor": "0.3-0.5",
            "manufacturer_data": {
                "standard": 0.4,
                "long_radius": 0.2,
                "short_radius": 0.5
            },
            "reference": "Crane Technical Paper No. 410"
        },
        {
            "type": "u_bend",
            "description": "U-bend or return bend",
            "typical_k_factor": "1.5-2.5",
            "manufacturer_data": {
                "standard": 2.0
            },
            "reference": "Crane Technical Paper No. 410"
        },
        {
            "type": "tee_elbow",
            "description": "Tee with elbow (branch flow)",
            "typical_k_factor": "1.0-2.0",
            "manufacturer_data": {
                "run_through": 0.4,
                "branch_flow": 1.5
            },
            "reference": "Crane Technical Paper No. 410"
        },
        {
            "type": "tee_through",
            "description": "Tee with straight run (through flow)",
            "typical_k_factor": "0.2-0.6",
            "manufacturer_data": {
                "run_through": 0.4,
                "branch_flow": 1.5
            },
            "reference": "Crane Technical Paper No. 410"
        },
        {
            "type": "block_valve_full_line_size",
            "description": "Block valve (full line size)",
            "typical_k_factor": "0.5-1.0",
            "manufacturer_data": {
                "ball_valve": 0.5,
                "gate_valve": 0.2,
                "globe_valve": 3.0
            },
            "reference": "Manufacturer data sheets"
        },
        {
            "type": "block_valve_reduced_trim_0.9d",
            "description": "Block valve (reduced trim, 90% diameter)",
            "typical_k_factor": "1.0-2.0",
            "manufacturer_data": {
                "reduced_90": 1.5
            },
            "reference": "Manufacturer data sheets"
        },
        {
            "type": "block_valve_reduced_trim_0.8d",
            "description": "Block valve (reduced trim, 80% diameter)",
            "typical_k_factor": "2.0-4.0",
            "manufacturer_data": {
                "reduced_80": 3.0
            },
            "reference": "Manufacturer data sheets"
        },
        {
            "type": "globe_valve",
            "description": "Globe valve",
            "typical_k_factor": "6.0-10.0",
            "manufacturer_data": {
                "standard": 8.0,
                "high_performance": 4.0
            },
            "reference": "Manufacturer data sheets"
        },
        {
            "type": "diaphragm_valve",
            "description": "Diaphragm valve",
            "typical_k_factor": "2.0-5.0",
            "manufacturer_data": {
                "standard": 3.5
            },
            "reference": "Manufacturer data sheets"
        },
        {
            "type": "butterfly_valve",
            "description": "Butterfly valve",
            "typical_k_factor": "0.5-1.5",
            "manufacturer_data": {
                "standard": 1.0,
                "high_performance": 0.5
            },
            "reference": "Manufacturer data sheets"
        },
        {
            "type": "check_valve_swing",
            "description": "Swing check valve",
            "typical_k_factor": "2.0-5.0",
            "manufacturer_data": {
                "standard": 3.5
            },
            "reference": "Manufacturer data sheets"
        },
        {
            "type": "lift_check_valve",
            "description": "Lift check valve",
            "typical_k_factor": "3.0-6.0",
            "manufacturer_data": {
                "standard": 4.5
            },
            "reference": "Manufacturer data sheets"
        },
        {
            "type": "tilting_check_valve",
            "description": "Tilting disc check valve",
            "typical_k_factor": "1.0-2.0",
            "manufacturer_data": {
                "standard": 1.5
            },
            "reference": "Manufacturer data sheets"
        },
        {
            "type": "pipe_entrance_normal",
            "description": "Pipe entrance (sharp edge)",
            "typical_k_factor": "0.5",
            "manufacturer_data": {
                "sharp_edge": 0.5,
                "rounded": 0.04
            },
            "reference": "Crane Technical Paper No. 410"
        },
        {
            "type": "pipe_entrance_raise",
            "description": "Raised pipe entrance",
            "typical_k_factor": "0.2-0.5",
            "manufacturer_data": {
                "standard": 0.35
            },
            "reference": "Crane Technical Paper No. 410"
        },
        {
            "type": "pipe_exit",
            "description": "Pipe exit (discharge)",
            "typical_k_factor": "1.0",
            "manufacturer_data": {
                "standard": 1.0
            },
            "reference": "Crane Technical Paper No. 410"
        },
        {
            "type": "inlet_swage",
            "description": "Inlet swage (diameter transition)",
            "typical_k_factor": "0.1-0.5",
            "manufacturer_data": {
                "concentric": 0.2,
                "eccentric": 0.3
            },
            "reference": "Process piping design standards"
        },
        {
            "type": "outlet_swage",
            "description": "Outlet swage (diameter transition)",
            "typical_k_factor": "0.1-0.5",
            "manufacturer_data": {
                "concentric": 0.2,
                "eccentric": 0.3
            },
            "reference": "Process piping design standards"
        }
    ]


def get_template_categories() -> List[Dict[str, str]]:
    """Get available template categories.
    
    Returns:
        List of category dictionaries
    """
    return [
        {
            "id": "liquid",
            "name": "Liquid Systems",
            "description": "Templates for liquid flow calculations including water, oil, and other liquids"
        },
        {
            "id": "gas",
            "name": "Gas Systems",
            "description": "Templates for gas flow calculations including natural gas, air, and other gases"
        },
        {
            "id": "vapor",
            "name": "Vapor Systems",
            "description": "Templates for vapor/steam flow calculations"
        },
        {
            "id": "simple",
            "name": "Simple Examples",
            "description": "Basic templates for learning and testing"
        },
        {
            "id": "industrial",
            "name": "Industrial Applications",
            "description": "Templates for common industrial applications"
        }
    ]


# Template registry
TEMPLATES = {
    "liquid_system": {
        "name": "Liquid System Example",
        "description": "Complete liquid system with multiple sections",
        "category": "liquid",
        "config": get_liquid_system_template()
    },
    "gas_system": {
        "name": "Gas System Example",
        "description": "Complete gas system with control valve",
        "category": "gas",
        "config": get_gas_system_template()
    },
    "vapor_system": {
        "name": "Steam System Example",
        "description": "Steam system with adiabatic flow",
        "category": "vapor",
        "config": get_vapor_system_template()
    },
    "simple_liquid": {
        "name": "Simple Liquid Test",
        "description": "Simple liquid system for testing",
        "category": "simple",
        "config": get_simple_liquid_template()
    }
}


def get_template(template_id: str) -> Dict[str, Any]:
    """Get a specific template by ID.
    
    Args:
        template_id: Template identifier
        
    Returns:
        Template configuration dictionary
        
    Raises:
        ValueError: If template not found
    """
    if template_id not in TEMPLATES:
        available_templates = list(TEMPLATES.keys())
        raise ValueError(f"Template '{template_id}' not found. Available templates: {available_templates}")
    
    return TEMPLATES[template_id]["config"]


def list_templates() -> List[Dict[str, Any]]:
    """List all available templates.
    
    Returns:
        List of template metadata
    """
    return [
        {
            "id": template_id,
            "name": template_data["name"],
            "description": template_data["description"],
            "category": template_data["category"]
        }
        for template_id, template_data in TEMPLATES.items()
    ]


def list_templates_by_category(category: str) -> List[Dict[str, Any]]:
    """List templates filtered by category.
    
    Args:
        category: Template category
        
    Returns:
        List of template metadata for the category
    """
    return [
        {
            "id": template_id,
            "name": template_data["name"],
            "description": template_data["description"],
            "category": template_data["category"]
        }
        for template_id, template_data in TEMPLATES.items()
        if template_data["category"] == category
    ]